lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,147 @@ namespace Js
+    Var JavascriptPromise::EntryThenFinallyFunction(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        ScriptContext* scriptContext = function->GetScriptContext();
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        JavascriptPromiseThenFinallyFunction* This = JavascriptPromiseThenFinallyFunction::FromVar(function);
+
+        //1. Let onFinally be F.[[OnFinally]]
+        //2. Assert: IsCallabale(onFinally)=true
+        Assert(JavascriptConversion::IsCallable(This->GetOnFinally()));
+
+        //3. Let result be ? Call(onFinally, undefined)
+        Var result = CALL_FUNCTION(scriptContext->GetThreadContext(), This->GetOnFinally(), CallInfo(CallFlags_Value, 1), library->GetUndefined());
+
+        //4. Let C be F.[[Constructor]]
+        //5.Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(This->GetConstructor()));
+
+        //6. Let promise be ? PromiseResolve(c, result)
+        Var promiseVar = CreateResolvedPromise(result, scriptContext, This->GetConstructor());
+
+        //7. Let valueThunk be equivalent to a function that returns value
+        //OR 7. Let thrower be equivalent to a function that throws reason
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
+
+        //8. Return ? Invoke(promise, "then", <<valueThunk>>)
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(promiseVar);
+        Var funcVar = JavascriptOperators::GetProperty(promise, Js::PropertyIds::then, scriptContext);
+
+        if (!JavascriptConversion::IsCallable(funcVar))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction, _u("Promise.prototype.finally"));
+        }
+
+        RecyclableObject* func = RecyclableObject::FromVar(funcVar);
+        RecyclableObject* undefinedVar = scriptContext->GetLibrary()->GetUndefined();
+
+        return CALL_FUNCTION(scriptContext->GetThreadContext(),
+            func, Js::CallInfo(CallFlags_Value, 3),
+            promiseVar,
+            Thunk,
+            undefinedVar);
+    }
+
+    // valueThunk Function as described in draft ES 2018Section 25.4.5.3.1.7
+    // and thrower as described in draft ES 2018Section 25.4.5.3.2.7
+    Var JavascriptPromise::EntryThunkFinallyFunction(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        JavascriptPromiseThunkFinallyFunction* This = JavascriptPromiseThunkFinallyFunction::FromVar(function);
```

Comment:
One more nitpick about `This`, can we name it something like `thunkFinallyFunction`?

Reply:
changed.